src/main.cpp

```diff
@@ -2479,7 +2479,7 @@ bool DisconnectTip(CValidationState &state, const Consensus::Params &consensusPa
     int64_t nStart = GetTimeMicros();
     {
         CCoinsViewCache view(pcoinsTip);
-        if (DisconnectBlock(block, pindexDelete, view) != DISCONNECT_OK)
+        if (DisconnectBlock(MakeBlockRef(block), pindexDelete, view) != DISCONNECT_OK)
```

Comment:
At this abstraction level, it looks like MakeBlockRef is going to make a copy of block on the heap and then hook a shared pointer to it.  Since the block is already allocated on the stack, this looks to be very inefficient.  But the magic of std::forward and c++11's && parameter probably removes most of this inefficiency, but there is a hidden gotcha which is that "block" can't be used after the call to MakeBlockRef().  

However, if the "CBlock block" was converted to a smart pointer (so no stack object was ever used), this code would be a lot cleaner:
```
CBlockRef block;
if (!ReadBlockFromDisk(block, pindexDelete, consensusParams))
         return AbortNode(state, "Failed to read block");
      // Apply the block atomically to the chain state.
      int64_t nStart = GetTimeMicros();
     {
       CCoinsViewCache view(pcoinsTip);
       if (DisconnectBlock(block, pindexDelete, view) != DISCONNECT_OK)
```

Reply:
... perhaps ReadBlockFromDisk could return a CBlockRef?

Reply:
I don't see a performance issue here but as you say there's a gotcha and we'd have to be careful not so use block after we create the CBlockRef, but yeah, I'm working toward getting it done all the way through to read and write and also in the mining side of things.